### PR TITLE
Android CI Testing

### PR DESCRIPTION
**Pull request by sbSteveK**
* Setup android CI Testing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: utils/run_android_ci.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    with open(build_file_location, 'rb') as f:
+        data = f.read()
+    r = requests.put(device_farm_upload_url, data=data)
+    print('File upload status code: ' + str(r.status_code) + ' reason: ' + r.reason)
+
+    device_farm_upload_status = client.get_upload(arn=device_farm_upload_arn)
+    while device_farm_upload_status['upload']['status'] != 'SUCCEEDED':
+        if device_farm_upload_status['upload']['status'] == 'FAILED':
+            print('Upload failed to process')
+            exit -1
```
Comment (sfod) [inline utils/run_android_ci.py:65]: It's more a question than a comment: Why we use -1 as the exit code?

But, while trying to figure out how CI reports -1 exit codes, I found out that `exit -1` is not a proper syntax, it should be `exit(-1)`. 
But another but, according to https://docs.python.org/3/library/constants.html, `exit` shouldn't be used in programs. Instead, `sys.exit()` is the right choice

File: utils/run_android_ci.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    create_upload_response = client.create_upload(
+        projectArn=project_arn,
+        name=upload_test_file_name,
+        type='INSTRUMENTATION_TEST_PACKAGE'
+    )
+    device_farm_instrumentation_upload_arn = create_upload_response['upload']['arn']
+    device_farm_instrumentation_upload_url = create_upload_response['upload']['url']
+
+    with open(test_file_location, 'rb') as f:
+        dataInsturmentation = f.read()
```
Comment (sfod) [inline utils/run_android_ci.py:80]: Trivial: `dataInsturmentation` is camelCase, while other vars are snake_case

File: utils/run_android_ci.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    run_end_date_time = run_end_time.strftime("%m/%d/%Y, %H:%M:%S")
+    print('Run ended at ' + run_end_date_time + ' with result: ' + get_run_response['run']['result'])
+
+    is_success = True
+    if get_run_response['run']['result'] != 'PASSED':
+        print('run has failed with result ' + get_run_response['run']['result'])
+        is_success = False
+
+    # Clean up
+    print('Deleting' + upload_file_name + ' Device Farm project')
```
Comment (sfod) [inline utils/run_android_ci.py:129]: Trivial: Missing space between "Deleting" and `upload_file_name`

File: utils/run_android_ci.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    upload_file_name = 'CI-' + run_id + '-' + run_attempt + '.apk'
+    print('Upload file name: ' + upload_file_name)
+
+    # Setup upload to Device Farm project
+    create_upload_response = client.create_upload(
+        projectArn=project_arn,
+        name=upload_file_name,
+        type='ANDROID_APP'
+    )
+    device_farm_upload_arn = create_upload_response['upload']['arn']
```
Comment (sfod) [inline utils/run_android_ci.py:52]: Does `client.create_upload()` always succeed? For the requests below we check the `['upload']['status']` field

File: utils/run_android_ci.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    # Upload project apk
+    with open(build_file_location, 'rb') as f:
+        data = f.read()
+    r = requests.put(device_farm_upload_url, data=data)
+    print('File upload status code: ' + str(r.status_code) + ' reason: ' + r.reason)
+
+    device_farm_upload_status = client.get_upload(arn=device_farm_upload_arn)
+    while device_farm_upload_status['upload']['status'] != 'SUCCEEDED':
+        if device_farm_upload_status['upload']['status'] == 'FAILED':
```
Comment (sfod) [inline utils/run_android_ci.py:63]: Only `FAILED` and `SUCCEEDED` values are possible for the status? If other values are possible, this loop can become infinite

File: sdk/tests/android/testapp/src/androidTest/java/software/amazon/awssdk/iottest/MainActivityTest.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
+package software.amazon.awssdk.iottest
```
Comment (sfod) [inline sdk/tests/android/testapp/src/androidTest/java/software/amazon/awssdk/iottest/MainActivityTest.kt:1]: A missing copyright header? Not sure about this tbh

File: sdk/tests/android/testapp/src/main/java/software/amazon/awssdk/iottest/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package software.amazon.awssdk.iottest
+
+import androidx.appcompat.app.AppCompatActivity
+
+class MainActivity : AppCompatActivity() {
```
Comment (sfod) [inline sdk/tests/android/testapp/src/main/java/software/amazon/awssdk/iottest/MainActivity.kt:5]: What's the purpose of this file? If it's a stab for the future work, then we can keep it

**Review by sfod — APPROVED** [review]

File: utils/run_android_ci.py (hunk 1 above)
Comment (sbSteveK) [reply utils/run_android_ci.py:65]: I remember asking this a while back and vaguely recall it has something to do with indicating to CI that something has gone wrong and to fail the action/test.

yup, should be `sys.exit(-1)`

File: utils/run_android_ci.py (hunk 4 above)
Comment (sbSteveK) [reply utils/run_android_ci.py:52]: The `create_upload` should throw an exception if something is wrong with the request: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/devicefarm/client/create_upload.html

The `upload_status` check using the arn is checking the file's status before proceeding to make sure it's ready for use before the next step. The status in that check mirrors the status received in `create_upload` for that specific arn.

File: sdk/tests/android/testapp/src/androidTest/java/software/amazon/awssdk/iottest/MainActivityTest.kt (hunk 6 above)
Comment (sbSteveK) [reply sdk/tests/android/testapp/src/androidTest/java/software/amazon/awssdk/iottest/MainActivityTest.kt:1]: Adding them everywhere =P

File: sdk/tests/android/testapp/src/main/java/software/amazon/awssdk/iottest/MainActivity.kt (hunk 7 above)
Comment (sbSteveK) [reply sdk/tests/android/testapp/src/main/java/software/amazon/awssdk/iottest/MainActivity.kt:5]: It gets built and installed with everything included in its gradle.build file (the samples src files, the sdk, the crt library) and then the instrumented tests get run against it. We upload both the app package built from this as well as the app package built from the tests and run them in parallel.

**Pull request merged**